Allow Web Cmdlets to Ignore HTTP Error Statuses

vdamewood opened the pull request:
# PR Summary
Fix #5555
Fix #9769

This PR adds a  -SkipHttpErrorCheck flag to Invoke-WebRequest and Invoke-RestMethod, and adds a -ResponseStatusVariable option to Invoke-RestMethod. The -SkipHttpErrorCheck flag causes the cmdlets to ignore HTTP error statuses and continue to process the responses and write to the pipeline as if these had been successful. The -ResponseStatusVariable assigns the integer value of the status code to the variable named. It can be used with or without -SkipHttpErrorCheck.

## PR Context

These cmdlets as they are currently written will throw an exception when the HTTP status is not considered a success. The error handling code will extract the body of the response and modify its contents to assign to the exception, but will leave the original body contents inaccessible. This isn't feasible with web APIs and other sites that return useful diagnostic messages in the body that would be modified by the error handling code.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4724
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

vdamewood committed:
Add -SkipHttpErrorCheck to web request cmdlets

The -SkipHttpErrorCheck flag causes web request cmdlets
(Invoke-WebRequest and Invoke-RestMethod) to ignore HTTP statuses that
are error statuses and treat them as successful requests. This allows
users to handle the responses using their own error handler and gives
them access to the full, unmodified response body and headers.

vdamewood committed:
Add -ResponseStatusVariable to Invoke-RestMethod

This allows the user to specify a variable to set to the integer value
of the respons's status code, Analogous to using
-ResponseHeadersVariable to retrieve the headers of the response. This
can be used to distinguish error messages from success messages when
used with -SkipHttpErrorCheck.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=10466) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=10466) before we can accept your contribution.<br/><br/>:x: vdamewood [sign now](https://cla.opensource.microsoft.com/PowerShell/PowerShell?pullRequest=10466)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/PowerShell/PowerShell?pullRequest=10466) it.</sub>

vdamewood committed:
Fix coding-style error

The summary for the SkipHttpErrorCheck property didn't conform
to style guidelines. This changes the summary to start with
"Gets or sets".

markekraus reviewed (changes requested):
Changes look mostly good. Please add test coverage per conversation in PowerShell bridge.

vdamewood committed:
Add tests for -SkipHttpErrorCheck

This flag supresses terminating errors on web cmdlets.
The tests are written to check that it properly
supressed the errors.

vdamewood committed:
Add test for -StatusCodeVariable

Th -StatusCodeVariable parameter specifies an output
variable for the status code with Invoke-RestMethod.
This test makes sure it functions properly.

vdamewood commented:
I've added the tests, as requested.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,23 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        }
+
+        $Uri =  Get-WebListenerUrl -Test 'Response' -Query $Query
+        $command = "Invoke-WebRequest -SkipHttpErrorCheck -Uri '$uri'"
```
I'd expect a check if we do not have SkipHttpErrorCheck parameter to ensure that the parameter really works.

markekraus replied:
there are already tests for error handling that would have failed if ` -SkipHttpErrorCheck` was present. I think those are sufficient to test for and prevent regression.

annoying258 replied:
are all the relevant tests in https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1 (branch:`master`) ?

markekraus replied:
yes

iSazonov replied:
If there are tests for errors they is other test - they could be changed or removed. My request here is make the test more clear and reliable - that for the _same_ input/environment we have predictable behavior with and without the parameter.

vdamewood replied:
I agree with iSazonov. The other tests could be changed for whatever reason related to what they're meant to test. I've added failure tests.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,23 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+
+        $Uri =  Get-WebListenerUrl -Test 'Response' -Query $Query
+        $command = "Invoke-WebRequest -SkipHttpErrorCheck -Uri '$uri'"
+        $result = ExecuteWebCommand -command $command
```
Please start parameter names with upper case
```suggestion
        $result = ExecuteWebCommand -Command $command
```
Below too.
 Should -be ->  Should -Be 
and so on.

vdamewood replied:
Done. I've also fixed some inconsistent capitalization with variable names.

markekraus reviewed (approved):
Approved pending minor fixes suggested by @iSazonov

annoying258 commented:
its easy to understand what is going on, because of the really high-quality commit-messages and descriptions. ＼(＾▽＾)／

there are some typos. all the suggested changes are in commit descriptions.
this definitely needs automation for simple typos.

format for the changes is :

commit-hash

before 
`after`

---

7a85e0f6a4c6788a74566e841b7cdefe7068196a 


unmodified response body
`unmodified response-body`

---

53ec3c5258450f941e6ff81d26abe9aa1b96c4a5 

Analogous to
`analogous to`

---

b1e43e12a6a64ce1bb10d709bfa400b0a8ab124f

Th -StatusCodeVariable
`The -StatusCodeVariable`

---

annoying258 commented:
another suggestion, same format

**improving variable names**:

https://github.com/PowerShell/PowerShell/pull/10466/files/b1e43e12a6a64ce1bb10d709bfa400b0a8ab124f#diff-c638a7a1c62868071f9b1ec9c8fdfcc6R788

body
`responsebody`

---


https://github.com/PowerShell/PowerShell/pull/10466/files/b1e43e12a6a64ce1bb10d709bfa400b0a8ab124f#diff-c638a7a1c62868071f9b1ec9c8fdfcc6R789

headers
`responseheaders`

vdamewood committed:
Fix typos and style for -SkipHttpErrorCheck tests

Variables had different cases from each other and some parameter
names were lower case.

vdamewood committed:
Fix typos and style for -StatusCodeVariable test

Variables had different cases from each other and some parameter
names were lower case.

vdamewood committed:
Add failure tests when missing -SkipHttpErrorCheck

These tests ensure that Web Cmdlets fail when -SkipHttpErrorCheck
is missing.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,39 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $result = ExecuteWebCommand -Command $command
+        $result.output.StatusCode | Should -Be 404
+        $result.output.Content | Should -Match "oops"
+        $result.error | Should -Be $null
```
```suggestion
        $result.error | Should -BeNullOrEmpty
```

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,39 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $uri =  Get-WebListenerUrl -Test 'Response' -Query $query
+        $command = "Invoke-WebRequest -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output | Should -Be $null
```
```suggestion
        $result.output | Should -BeNullOrEmpty
```

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,39 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $command = "Invoke-WebRequest -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output | Should -Be $null
+        $result.error | Should -Not -Be $null
```
```suggestion
        $result.error | Should -Not -BeNullOrEmpty
```

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2259,6 +2292,52 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $command = "Invoke-RestMethod -SkipHttpErrorCheck -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output.message | Should -Match "oops"
+        $result.output.error | Should -Be $null
```
```suggestion
        $result.output.error | Should -BeNullOrEmpty
```

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -780,6 +780,39 @@ Describe "Invoke-WebRequest tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $command = "Invoke-WebRequest -SkipHttpErrorCheck -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output.StatusCode | Should -Be 404
+        $result.output.Content | Should -Match "oops"
```
```suggestion
        $result.output.Content | Should -BeExactly "oops"
```

`match` is used for regex matches.

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2259,6 +2292,52 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
+    }
+
+    It "Verify Invoke-RestMethod terminates without -SkipHttpErrorCheck" {
+        $query = @{
```
This query seems identical to the previous test. Please move this to `BeforeAll` block for this context.

vdamewood replied:
Done. I've also removed a similar redundancy in the tests for Invoke-WebRequest.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2259,6 +2292,52 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $uri =  Get-WebListenerUrl -Test 'Response' -Query $query
+        $command = "Invoke-RestMethod -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output | Should -Be $null
```
same as above

vdamewood replied:
Done.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2259,6 +2292,52 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
+        $command = "Invoke-RestMethod -Uri '$uri'"
+        $result = ExecuteWebCommand -Command $command
+        $result.output | Should -Be $null
+        $result.error | Should -Not -Be $null
```
same as above

vdamewood replied:
Done.

vdamewood committed:
Clean up tests for -SkipHttpErrorCheck

Per discussion on the pull requests. This commit fixes up
style problems with the tests for -SkipHttpErrorCheck.

vdamewood committed:
Add more status tests for -StatusCodeVariable

Previously, the -StatusCodeVariable flag in Invoke-RestMethod
only had tests for 200 status. This commit adds tests for 404
and 500 statuses and removes -SkipHttpErrorCheck from the 200
check.

vdamewood commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -2259,6 +2292,52 @@ Describe "Invoke-RestMethod tests" -Tags "Feature", "RequireAdminOnWindows" {
+
+        $uri =  Get-WebListenerUrl -Test 'Response' -Query $query
+        Invoke-RestMethod -SkipHttpErrorCheck -StatusCodeVariable code -Uri "$uri"
+        $code | Should -Be 200
```
I've added a 404 and a 500 test. Let me know if you meant 400 specifically and not just any 4XX-range status.

vdamewood committed:
Fix response body in -StatusCodeVariable test

The test was copy/pastes from the 200 status test. The body
indicated success. This commit fixes that so it is also an
error indicator.

vdamewood commented:
Is there anything else I need to do to get this done?

adityapatwardhan commented:
@TravisEz13 I have signed off, sorry for the delay.

iSazonov commented:
@vdamewood Thanks for your contribution!
